app/src/main/kotlin/viewmodel/graph/EdgeViewModel.kt

```diff
@@ -0,0 +1,77 @@
+package viewmodel.graph
+
+import androidx.compose.runtime.State
+import androidx.compose.ui.unit.Dp
+import ARROW_DEPTH
+import ARROW_SIZE
+import SQRT_3
+import androidx.compose.runtime.mutableStateOf
+import androidx.compose.ui.graphics.Color
+import kotlin.math.sqrt
+
+class EdgeViewModel<D>(
+    val firstVertex: VertexViewModel<D>,
+    val secondVertex: VertexViewModel<D>,
+    private val isDirected: State<Boolean>
+) {
+
+    fun isDirected() = isDirected.value
+
+    private val radius = firstVertex.radius
+
+    var highlightColor = mutableStateOf(Color.LightGray)
+
+    internal fun calculateFirstVertexCenter(scale: Float): Pair<Dp, Dp> {
+        val x = firstVertex.x.value + radius * scale
+        val y = firstVertex.y.value + radius * scale
+
+        return Pair(x, y)
+    }
+
+    internal fun calculateSecondVertexCenter(scale: Float): Pair<Dp, Dp> {
+        val x = secondVertex.x.value + radius * scale
+        val y = secondVertex.y.value + radius * scale
+
+        return Pair(x, y)
+    }
+
+    internal fun calculateArrowPoints(scale: Float): List<Pair<Dp, Dp>> {
+        if (!isDirected.value) return listOf()
+
+        val firstVertexCenterX = calculateFirstVertexCenter(scale).first
+        val firstVertexCenterY = calculateFirstVertexCenter(scale).second
+
+        val secondVertexCenterX = calculateSecondVertexCenter(scale).first
+        val secondVertexCenterY = calculateSecondVertexCenter(scale).second
+
+        val vectorX = secondVertexCenterX - firstVertexCenterX
+        val vectorY = secondVertexCenterY - firstVertexCenterY
+
+        val len = sqrt(vectorX.value * vectorX.value + vectorY.value * vectorY.value)
+        val normedVectorX = vectorX / len
+        val normedVectorY = vectorY / len
+
+        // rotate normed vector by Pi/6
+        val aX = normedVectorX * SQRT_3 / 2 - normedVectorY * 1 / 2
+        val aY = normedVectorX * 1 / 2 + normedVectorY * SQRT_3 / 2
+
+        // rotate normed vector by negative Pi/6
+        val bX = normedVectorX * SQRT_3 / 2 + normedVectorY * 1 / 2
+        val bY = -normedVectorX * 1 / 2 + normedVectorY * SQRT_3 / 2
+
+        val arrowEndPointX = secondVertexCenterX - normedVectorX * (radius.value - ARROW_DEPTH) * scale
+        val arrowEndPointY = secondVertexCenterY - normedVectorY * (radius.value - ARROW_DEPTH) * scale
+
+        val arrowLeftPointX = arrowEndPointX - aX * ARROW_SIZE * scale
+        val arrowLeftPointY = arrowEndPointY - aY * ARROW_SIZE * scale
+
+        val arrowRightPointX = arrowEndPointX - bX * ARROW_SIZE * scale
+        val arrowRightPointY = arrowEndPointY - bY * ARROW_SIZE * scale
+
+        return listOf(
+            Pair(arrowEndPointX, arrowEndPointY),
+            Pair(arrowLeftPointX, arrowLeftPointY),
+            Pair(arrowRightPointX, arrowRightPointY)
+        )
+    }
+}
```

Comment:
Thanks for your suggestion, may I ask where can we calculate these points instead? Don't we need to find them whenever a vertex with edges is moved, so arrows can be correctly displayed?

Reply:
I actually only thought about the case when the whole graph is moved. With individual movements, yes, it seems like you'll need to do the calculations individually.

But my main point (which I didn't state too clearly, I apologize) was actually that in MVVM all this should work like so: a vertex gets dragged ->it calls a method of the view model to tell it how much it needs to be dragged -> view model does the calculations and saves them in its mutable state property -> the view sees that the property has changed and automatically gets recomposed. So, instead of returning the result it should be saved in a state that the view would react to.